browse/controllers/archive_page/__init__.py

```diff
@@ -0,0 +1,140 @@
+"""Archive landing page."""
+
+from datetime import date
+from typing import Dict, Any, Tuple, List
+
+from flask import Response, url_for
+
+from arxiv import status
+from arxiv.taxonomy.definitions import ARCHIVES, CATEGORIES, ARCHIVES_SUBSUMED
+
+from browse.controllers.archive_page.by_month_form import ByMonthForm
+from browse.controllers.archive_page.catchup_form import CatchupForm
+from browse.services.util.response_headers import abs_expires_header
+
+
+def get_archive(archive_id: str) -> Response:
+    """Gets archive page."""
+    data: Dict[str, Any] = {}
+    response_headers: Dict[str, Any] = {}
+
+    if archive_id == "list":
+        return archive_index(archive_id, status=status.HTTP_200_OK)
+
+    archive = ARCHIVES.get(archive_id, None)
+    if not archive:
+        cat_id = CATEGORIES.get(archive_id, {}).get("in_archive", None)
+        archive = ARCHIVES.get(cat_id, None)
+        if not archive:
+            return archive_index(archive_id,
+                                 status=status.HTTP_404_NOT_FOUND)
+        else:
+            archive_id = cat_id
+
+    _write_expires_header(response_headers)
+
+    subsumed_by = ARCHIVES_SUBSUMED.get(archive_id, None)
+    if subsumed_by:
+        data["subsumed_id"] = archive_id
+        data["subsumed_category"] = CATEGORIES.get(archive_id, {})
+        data["subsumed_by"] = subsumed_by
+        subsuming_category = CATEGORIES.get(subsumed_by, {})
+        data["subsuming_category"] = subsuming_category
+        archive_id = subsuming_category.get("in_archive", None)
+        archive = ARCHIVES.get(archive_id, None)
+
+    years = years_operating(archive)
+
+    data["archive_id"] = archive_id
+    data["archive"] = archive
+    data["list_form"] = ByMonthForm(archive_id, archive, years)
+    data["catchup_form"] = CatchupForm(archive_id, archive, years)
+    data["stats_by_year"] = stats_by_year(archive_id, archive, years)
+    data["category_list"] = category_list(archive_id)
+
+    data["template"] = "archive/single_archive.html"
+    return data, status.HTTP_200_OK, response_headers  # type: ignore
+
+
+def archive_index(archive_id: str, status: int) -> Response:
+    """Landing page for when there is no archive specified."""
+    data: Dict[str, Any] = {}
+    data["bad_archive"] = archive_id
+
+    archives = [
+        (id, ARCHIVES[id]["name"])
+        for id in ARCHIVES.keys()
+        if id not in ARCHIVES_SUBSUMED and not id.startswith("test")
+    ]
```

Comment:
Just to confirm I know what is happening here, I think Python is flattening the elements of what is inside the `[` `]` to be a list. But it looks like this almost a reverse-map style of syntax, where the output is specified first (`(id, ARCHIVES[id]["name"])`) then the list of things to map over is specified (the filtered `id`s). Would be good to know what this is. Also, are `archives` just `List[str, str]`? might be nice to annotate to make it a little more clear

Reply:
You have it. The reverse-map syntax is python list comprehension. 
I often use flatten (as in flatmap) to mean [[a],[b,c]] => [a,b,c] which isn't happening here. But I think you mean flaten in a metaphorical sense.